Implemented Armeria-based http client binding for ScribeJava.

max904-github opened the pull request:
[Armeria](https://github.com/line/armeria/) is an open-source
asynchronous HTTP/2/1 client/server library built on top of [Netty](https://github.com/netty/netty)
and Java 8. Java 8 required to build this HTTP client module.

kullfar commented:
Hi!
First of all, thanks for the great Pull Request! Second, sorry for a long delay with the answer.

Next. I have tried to run the Example using Ameria and got an error.

I took https://github.com/scribejava/scribejava/blob/master/scribejava-apis/src/test/java/com/github/scribejava/apis/examples/Google20Example.java

and just add ```.httpClientConfig(ArmeriaHttpClientConfig.defaultConfig())```

And I got this error
```Exception in thread "main" com.github.scribejava.core.model.OAuth2AccessTokenErrorResponse: {
  "error": {
    "code": 400,
    "message": "Invalid JSON payload received. Unexpected token.\ncode=4%2FzgEhHDVnkft\n^",
    "status": "INVALID_ARGUMENT"
  }
}```

It seems, not all data are sent to the google server.
I've tried the latest ameria and netty versions as well, no changes

kullfar commented on scribejava-httpclient-armeria/src/test/resources/simplelogger.properties:
```diff
@@ -0,0 +1,29 @@
+# SLF4J's SimpleLogger configuration file
+# Simple implementation of Logger that sends all enabled log messages, for all defined loggers, to System.err.
```
Does it really needed?

max904-github replied:
No, it's not necessary. I'll remove it with an updated PR.

max904-github commented:
> Hi!
> First of all, thanks for the great Pull Request! Second, sorry for a long delay with the answer.
> 
> Next. I have tried to run the Example using Ameria and got an error.
> 
> I took https://github.com/scribejava/scribejava/blob/master/scribejava-apis/src/test/java/com/github/scribejava/apis/examples/Google20Example.java
> 
> and just add `.httpClientConfig(ArmeriaHttpClientConfig.defaultConfig())`
> 
> And I got this error
> 
> ```
>   "error": {
>     "code": 400,
>     "message": "Invalid JSON payload received. Unexpected token.\ncode=4%2FzgEhHDVnkft\n^",
>     "status": "INVALID_ARGUMENT"
>   }
> }```
> 
> It seems, not all data are sent to the google server.
> I've tried the latest ameria and netty versions as well, no changes
> ```

Thank you for the feedback! I'll upgrade the Armeria version to the current and look into this problem.

max904-github commented:
I have upgraded Armeria to latest version 0.99.5 and made a number of enhancements, including:
- reworked ArmeriaHttpClientConfig and added more configuration options;
- added advanced retry calability;
- removed simplelogger.properties used for tests, as requested by @kullfar
- tested implementation using unit tests and GitHubExample

max904-github commented:
> Hi!
> First of all, thanks for the great Pull Request! Second, sorry for a long delay with the answer.
> 
> Next. I have tried to run the Example using Ameria and got an error.
> 
> I took https://github.com/scribejava/scribejava/blob/master/scribejava-apis/src/test/java/com/github/scribejava/apis/examples/Google20Example.java
> 
> and just add `.httpClientConfig(ArmeriaHttpClientConfig.defaultConfig())`
> 
> And I got this error
> 
> ```
>   "error": {
>     "code": 400,
>     "message": "Invalid JSON payload received. Unexpected token.\ncode=4%2FzgEhHDVnkft\n^",
>     "status": "INVALID_ARGUMENT"
>   }
> }```
> 
> It seems, not all data are sent to the google server.
> I've tried the latest ameria and netty versions as well, no changes
> ```

Unfortunately, I don't have Google Dev account to test with, but I have successfully tested with GitHub using GitHubExample from Scribejava. Here is the trace of my test:
```
=== GitHub's OAuth Workflow ===

Fetching the Authorization URL...
Got the Authorization URL!
Now go and authorize ScribeJava here:
https://github.com/login/oauth/authorize?response_type=code&client_id=***&redirect_uri=http%3A%2F%2Fwww.example.com%2Foauth_callback%2F&state=secret305415
And paste the authorization code here
>>***

And paste the state from server here. We have set 'secretState'='secret305415'.
>>secret305415
State value does match!
Trading the Authorization Code for an Access Token...
[16:19:10] [main] DEBUG SystemInfo - Java version: 11
[16:19:11] [main] DEBUG ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@32a68f4f
[16:19:11] [main] INFO JavaVersionSpecific - Using the APIs optimized for: Java 9+
[16:19:12] [main] DEBUG DefaultDnsNameResolver - [github.com] Sending a DNS query: DnsQuestion(github.com IN A)
[16:19:12] [armeria-common-worker-nio-2-1] DEBUG SslHandler - [id: 0x5dc9f9e6, L:/192.168.2.13:63872 - R:github.com/140.82.118.3:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_128_GCM_SHA256
[16:19:12] [armeria-common-worker-nio-2-1] INFO LoggingClient - [creqId=07947c21, chanId=5dc9f9e6, laddr=192.168.2.13:63872, raddr=github.com/140.82.118.3:443][h1://github.com/login/oauth/access_token#POST] Request: {startTime=2020-05-18T14:19:11.915Z(1589811551915491), length=117B, duration=1007ms(1007865200ns), scheme=none+h1, headers=[:method=POST, :path=/login/oauth/access_token, :authority=github.com, :scheme=https, authorization=Basic ***, content-length=117]}
[16:19:13] [armeria-common-worker-nio-2-1] INFO LoggingClient - [creqId=07947c21, chanId=5dc9f9e6, laddr=192.168.2.13:63872, raddr=github.com/140.82.118.3:443][h1://github.com/login/oauth/access_token#POST] Response: {startTime=2020-05-18T14:19:13.482Z(1589811553482421), length=78B, duration=1792µs(1792400ns), totalDuration=1575ms(1575862400ns), headers=[:status=200, date=Mon, 18 May 2020 14:19:13 GMT, content-type=application/x-www-form-urlencoded; charset=utf-8, server=GitHub.com, status=200 OK, ...}
[16:19:13] [main] DEBUG DefaultDnsNameResolver - [api.github.com] Sending a DNS query: DnsQuestion(api.github.com IN A)
[16:19:13] [main] DEBUG DnsQueryContext - [id: 0x0561dca5] WRITE: UDP, [44292: /192.168.2.1:53], DefaultDnsQuestion(api.github.com. IN A)
Got the Access Token!
(The raw response looks like this: access_token=***&scope=&token_type=bearer')

Now we're going to access a protected resource...
[16:19:13] [armeria-common-worker-nio-2-2] DEBUG SslHandler - [id: 0xfd6832cd, L:/192.168.2.13:63874 - R:api.github.com/140.82.118.6:443] HANDSHAKEN: protocol:TLSv1.3 cipher suite:TLS_AES_128_GCM_SHA256
[16:19:13] [armeria-common-worker-nio-2-2] INFO LoggingClient - [creqId=c0c9fdfe, chanId=fd6832cd, laddr=192.168.2.13:63874, raddr=api.github.com/140.82.118.6:443][h1://api.github.com/user/repos#GET] Request: {startTime=2020-05-18T14:19:13.498Z(1589811553498001), length=0B, duration=141ms(141620700ns), scheme=none+h1, headers=[:method=GET, :path=/user/repos, :authority=api.github.com, :scheme=https, authorization=Bearer ***]}
[16:19:13] [armeria-common-worker-nio-2-2] INFO LoggingClient - [creqId=c0c9fdfe, chanId=fd6832cd, laddr=192.168.2.13:63874, raddr=api.github.com/140.82.118.6:443][h1://api.github.com/user/repos#GET] Response: {startTime=2020-05-18T14:19:13.921Z(1589811553921665), length=10557B, duration=15875µs(15875700ns), totalDuration=444ms(444881300ns), headers=[:status=200, date=Mon, 18 May 2020 14:19:13 GMT, content-type=application/json; charset=utf-8, content-length=10557, server=GitHub.com, status=200 OK, ...}
[16:19:13] [Thread-0] DEBUG ClientFactory - Closing the default client factories
[16:19:13] [armeria-common-worker-nio-2-2] DEBUG ClientFactory - Closed the default client factories
Got it! Lets see what we found...

200
[Response content from GitHub...]

That's it man! Go ahead and build something awesome with ScribeJava! :)

Process finished with exit code 0
```
The unit tests are running fine too!

kullfar commented:
Hi. I have finally figured out the problem.

That's it
https://github.com/scribejava/scribejava/commit/eccd3dd165e28447b7b6f9c574c7ab2289c42617#diff-df14ae25323f3adacf401c4209123a1fR132-R135

Your implementation missed setting of Content-Type header. Google thought you send json and cannot parse code from the POST request's body. I have added this block of code.

And I have added unit tests for this. https://github.com/scribejava/scribejava/blob/master/scribejava-core/src/test/java/com/github/scribejava/core/AbstractClientTest.java#L95

ps. And one more change to your code. I removed configuration wrappers for Armeria http client like this https://github.com/scribejava/scribejava/commit/eccd3dd165e28447b7b6f9c574c7ab2289c42617#diff-b5056cbbb34d7dace9aa8b6708783c52L134 
It's not ScribeJava's business to make such wrappers and maintain them. Configuration of http clients should be made outside of ScribeJava, like here https://github.com/scribejava/scribejava/commit/eccd3dd165e28447b7b6f9c574c7ab2289c42617#diff-df97542a1bda5f4eda9c5f787357a05cR45-R50 
So ScribeJava will have minimal dependency on Armeria's implementations and will not need any changes in case of Armeria changes in their configuration way.

kullfar commented:
merged in the master. thx
